File: packages/trading/octobot_trading/exchanges/util/exchange_market_status_fixer.py
#  Drakkar-Software OctoBot-Trading
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import decimal
import math

import octobot_commons.logging as logging
from octobot_trading.enums import ExchangeConstantsMarketStatusColumns as Ecmsc
from octobot_trading.enums import ExchangeConstantsMarketStatusInfoColumns as Ecmsic


def is_ms_valid(value, zero_valid=False):
    if isinstance(value, str):
        try:
            value = float(value)
        except ValueError:
            return False
    return value is not None and value is not math.nan and (value >= 0 if zero_valid else value > 0)


def check_market_status_limits(market_limit):
    return all([check_market_status_values(market_limit[key].values())
                for key in market_limit])


def check_market_status_values(values, zero_valid=False):
    return all([is_ms_valid(value, zero_valid=zero_valid) for value in values])


def get_markets_limit(market_limit):
    return market_limit[Ecmsc.LIMITS_COST.value] if Ecmsc.LIMITS_COST.value in market_limit else None, \
           market_limit[Ecmsc.LIMITS_PRICE.value] if Ecmsc.LIMITS_PRICE.value in market_limit else None, \
           market_limit[Ecmsc.LIMITS_AMOUNT.value] if Ecmsc.LIMITS_AMOUNT.value in market_limit else None


def calculate_amounts(market_limit):
    limit_cost, limit_price, limit_amount = get_markets_limit(market_limit)

    if not is_ms_valid(limit_amount[Ecmsc.LIMITS_AMOUNT_MAX.value]) and \
        Ecmsc.LIMITS_COST_MAX.value in limit_cost and Ecmsc.LIMITS_PRICE_MAX.value in limit_price:
        if is_ms_valid(limit_cost[Ecmsc.LIMITS_COST_MAX.value]) \
                and is_ms_valid(limit_price[Ecmsc.LIMITS_PRICE_MAX.value]) \
                and limit_price[Ecmsc.LIMITS_PRICE_MAX.value] > 0:
            limit_amount[Ecmsc.LIMITS_AMOUNT_MAX.value] = limit_cost[Ecmsc.LIMITS_COST_MAX.value] / \
                                                          limit_price[Ecmsc.LIMITS_PRICE_MAX.value]

    if not is_ms_valid(limit_amount[Ecmsc.LIMITS_AMOUNT_MIN.value]) and \
        Ecmsc.LIMITS_COST_MIN.value in limit_cost and Ecmsc.LIMITS_PRICE_MIN.value in limit_price:
        if is_ms_valid(limit_cost[Ecmsc.LIMITS_COST_MIN.value]) \
                and is_ms_valid(limit_price[Ecmsc.LIMITS_PRICE_MIN.value]) \
                and limit_price[Ecmsc.LIMITS_PRICE_MIN.value] > 0:
            limit_amount[Ecmsc.LIMITS_AMOUNT_MIN.value] = limit_cost[Ecmsc.LIMITS_COST_MIN.value] / \
                                                          limit_price[Ecmsc.LIMITS_PRICE_MIN.value]


def calculate_costs(market_limit):
    limit_cost, limit_price, limit_amount = get_markets_limit(market_limit)

    if not is_ms_valid(limit_cost[Ecmsc.LIMITS_COST_MAX.value]) and \
        Ecmsc.LIMITS_AMOUNT_MAX.value in limit_amount and Ecmsc.LIMITS_PRICE_MAX.value in limit_price:
        if is_ms_valid(limit_amount[Ecmsc.LIMITS_AMOUNT_MAX.value]) \
                and is_ms_valid(limit_price[Ecmsc.LIMITS_PRICE_MAX.value]):
            limit_cost[Ecmsc.LIMITS_COST_MAX.value] = limit_amount[Ecmsc.LIMITS_AMOUNT_MAX.value] * \
                                                      limit_price[Ecmsc.LIMITS_PRICE_MAX.value]

    if not is_ms_valid(limit_cost[Ecmsc.LIMITS_COST_MIN.value]) and \
        Ecmsc.LIMITS_AMOUNT_MIN.value in limit_amount and Ecmsc.LIMITS_PRICE_MIN.value in limit_price:
        if is_ms_valid(limit_amount[Ecmsc.LIMITS_AMOUNT_MIN.value]) \
                and is_ms_valid(limit_price[Ecmsc.LIMITS_PRICE_MIN.value]):
            limit_cost[Ecmsc.LIMITS_COST_MIN.value] = limit_amount[Ecmsc.LIMITS_AMOUNT_MIN.value] * \
                                                      limit_price[Ecmsc.LIMITS_PRICE_MIN.value]


def update_prices(market_limit):
    _, limit_price, _ = get_markets_limit(market_limit)

    if not is_ms_valid(limit_price[Ecmsc.LIMITS_PRICE_MAX.value]):
        limit_price[Ecmsc.LIMITS_PRICE_MAX.value] = None

    if not is_ms_valid(limit_price[Ecmsc.LIMITS_PRICE_MIN.value]):
        limit_price[Ecmsc.LIMITS_PRICE_MIN.value] = None


def fix_market_status_limits_from_current_data(market_limit):
    # calculate cost
    if not (check_market_status_values(market_limit[Ecmsc.LIMITS_COST.value].values())):
        calculate_costs(market_limit)

    # calculate amounts
    if not (check_market_status_values(market_limit[Ecmsc.LIMITS_AMOUNT.value].values())):
        calculate_amounts(market_limit)

    # set price to None if missing
    if not (check_market_status_values(market_limit[Ecmsc.LIMITS_PRICE.value].values())):
        update_prices(market_limit)

    if not is_ms_valid(
            market_limit[Ecmsc.LIMITS_COST.value][Ecmsc.LIMITS_COST_MIN.value]):
        market_limit[Ecmsc.LIMITS_COST.value][Ecmsc.LIMITS_COST_MIN.value] = 0


class ExchangeMarketStatusFixer:
    # todo move to connector/ccxt
    LIMIT_PRICE_MULTIPLIER = 1000
    LIMIT_COST_MULTIPLIER = 1

    # calculated from popular exchanges
    LIMIT_AMOUNT_MAX_SUP_ATTENUATION = 8  # when log(price, 10) >= 0
    LIMIT_AMOUNT_MAX_MINUS_3_ATTENUATION = 8  # when log(price, 10) < 0
    LIMIT_AMOUNT_MIN_ATTENUATION = 3  # when log(price, 10) < 0
    LIMIT_AMOUNT_MIN_SUP_ATTENUATION = 1  # when log(price, 10) >= 0

    """
    Utility class that performs exchange_self.market_status fixes
    """

    def __init__(self, market_status, price_example=None):
        self.logger = logging.get_logger(self.__class__.__name__)
        self.market_status = market_status
        self.price_example = price_example

        if Ecmsc.INFO.value in self.market_status:
            self.market_status_specific = self.market_status[Ecmsc.INFO.value]
        else:
            self.market_status_specific = None

        self._fix_typing()
        self._fix_market_status_precision()
        self._fix_market_status_limits()

    def _convert_values_to_float(self, element, parent_keys, key_whitelist):
        try:
            for parent_key in parent_keys:
                element = element[parent_key]
        except KeyError:
            # no parent_key in element, nothing to convert
            return
        for key, val in element.items():
            if key in key_whitelist and isinstance(val, str):
                try:
                    element[key] = float(val)
                except ValueError:
                    self.logger.debug(f"Impossible to convert {val} to float in {key} market status. "
                                      f"Full market status: {self.market_status}")

    def _fix_typing(self):
        self._convert_values_to_float(
            self.market_status,
            [Ecmsc.PRECISION.value, ],
            [Ecmsc.PRECISION_AMOUNT.value, Ecmsc.PRECISION_PRICE.value],
        )
        self._convert_values_to_float(
            self.market_status,
            [Ecmsc.LIMITS.value, Ecmsc.LIMITS_COST.value],
            [Ecmsc.LIMITS_COST_MAX.value, Ecmsc.LIMITS_COST_MIN.value],
        )
        self._convert_values_to_float(
            self.market_status,
            [Ecmsc.LIMITS.value, Ecmsc.LIMITS_AMOUNT.value],
            [Ecmsc.LIMITS_AMOUNT_MAX.value, Ecmsc.LIMITS_AMOUNT_MIN.value],
        )
        self._convert_values_to_float(
            self.market_status,
            [Ecmsc.LIMITS.value, Ecmsc.LIMITS_PRICE.value],
            [Ecmsc.LIMITS_PRICE_MAX.value, Ecmsc.LIMITS_PRICE_MIN.value],
        )

    def _fix_market_status_precision(self):
        if Ecmsc.PRECISION.value not in self.market_status:
            self.market_status[Ecmsc.PRECISION.value] = {
                Ecmsc.PRECISION_AMOUNT.value: None,
                Ecmsc.PRECISION_PRICE.value: None,
            }

        market_precision = self.market_status[Ecmsc.PRECISION.value]

        if not check_market_status_values(
            (market_precision.get(Ecmsc.PRECISION_AMOUNT.value), market_precision.get(Ecmsc.PRECISION_PRICE.value)),
            zero_valid=True
         ):
            if self.price_example is not None:
                self._fix_market_status_precision_with_price()

            elif self.market_status_specific:
                self._fix_market_status_precision_with_specific()

    def _fix_market_status_limits(self):
        if Ecmsc.LIMITS.value not in self.market_status:
            self.market_status[Ecmsc.LIMITS.value] = {}

        market_limit = self.market_status[Ecmsc.LIMITS.value]

        if Ecmsc.LIMITS_COST.value not in market_limit:
            market_limit[Ecmsc.LIMITS_COST.value] = {
                Ecmsc.LIMITS_COST_MAX.value: None,
                Ecmsc.LIMITS_COST_MIN.value: None
            }

        if Ecmsc.LIMITS_AMOUNT.value not in market_limit:
            market_limit[Ecmsc.LIMITS_AMOUNT.value] = {
                Ecmsc.LIMITS_AMOUNT_MAX.value: None,
                Ecmsc.LIMITS_AMOUNT_MIN.value: None
            }

        if Ecmsc.LIMITS_PRICE.value not in market_limit:
            market_limit[Ecmsc.LIMITS_PRICE.value] = {
                Ecmsc.LIMITS_PRICE_MAX.value: None,
                Ecmsc.LIMITS_PRICE_MIN.value: None
            }

        # if some data is missing
        if not check_market_status_limits(market_limit):
            fix_market_status_limits_from_current_data(market_limit)

            if self.market_status_specific and not check_market_status_limits(market_limit):
                self._fix_market_status_limits_with_specific()

            if self.price_example is not None and not check_market_status_limits(market_limit):
                self._fix_market_status_limits_with_price()

    def _calculate_amount(self):
        amount_log_price = math.log(self.price_example, 10)

        if amount_log_price >= 0:
            amount_min = 10 ** (self.LIMIT_AMOUNT_MIN_SUP_ATTENUATION - amount_log_price)
            amount_max = 10 ** (self.LIMIT_AMOUNT_MAX_SUP_ATTENUATION - amount_log_price)
        else:
            amount_min = 10 ** -(amount_log_price + self.LIMIT_AMOUNT_MIN_ATTENUATION)
            amount_max = 10 ** (-amount_log_price + self.LIMIT_AMOUNT_MAX_MINUS_3_ATTENUATION)

        return amount_min, amount_max

    def _fix_market_status_limits_with_price(self):
        candidate_amount_min, candidate_amount_max = self._calculate_amount()
        limits_amount = self.market_status[Ecmsc.LIMITS.value][Ecmsc.LIMITS_AMOUNT.value]
        amount_min = limits_amount.get(Ecmsc.LIMITS_AMOUNT_MIN.value)
        if not is_ms_valid(amount_min, zero_valid=True):
            amount_min = candidate_amount_min
        amount_max = limits_amount.get(Ecmsc.LIMITS_AMOUNT_MAX.value)
        if not is_ms_valid(amount_max, zero_valid=False):
            amount_max = candidate_amount_max

        limits_price = self.market_status[Ecmsc.LIMITS.value][Ecmsc.LIMITS_PRICE.value]
        price_min = limits_price.get(Ecmsc.LIMITS_PRICE_MIN.value)
        if not is_ms_valid(price_min, zero_valid=True):
            price_min = self.price_example / self.LIMIT_PRICE_MULTIPLIER
        price_max = limits_price.get(Ecmsc.LIMITS_PRICE_MAX.value)
        if not is_ms_valid(price_max, zero_valid=False):
            price_max = self.price_example * self.LIMIT_PRICE_MULTIPLIER
        limit_cost = self.market_status[Ecmsc.LIMITS.value].get(Ecmsc.LIMITS_COST.value, {})
        cost_min = limit_cost.get(Ecmsc.LIMITS_COST_MIN.value)
        if not is_ms_valid(cost_min, zero_valid=False):
            updated_cost = False
            if (
                is_ms_valid(limits_price.get(Ecmsc.LIMITS_PRICE_MIN.value))
                and is_ms_valid(limits_amount.get(Ecmsc.LIMITS_AMOUNT_MIN.value))
            ):
                # min cost can be computed from min price and amount: those values come from exchange data
                candidate_cost_min = (
                    limits_price.get(Ecmsc.LIMITS_PRICE_MIN.value)
                    * limits_amount.get(Ecmsc.LIMITS_AMOUNT_MIN.value)
                )
                if is_ms_valid(candidate_cost_min, zero_valid=False):
                    cost_min = candidate_cost_min
                    updated_cost = True
            if not updated_cost:
                # avoid computing min cost based on indirect min price or amount
                cost_min = 0
        cost_max = limit_cost.get(Ecmsc.LIMITS_COST_MAX.value)
        if not is_ms_valid(cost_max, zero_valid=False):
            cost_max = price_max * amount_max
        self.market_status[Ecmsc.LIMITS.value] = {
            Ecmsc.LIMITS_AMOUNT.value: {
                Ecmsc.LIMITS_AMOUNT_MIN.value: amount_min,
                Ecmsc.LIMITS_AMOUNT_MAX.value: amount_max,
            },
            Ecmsc.LIMITS_PRICE.value: {
                Ecmsc.LIMITS_PRICE_MIN.value: price_min,
                Ecmsc.LIMITS_PRICE_MAX.value: price_max,
            },
            Ecmsc.LIMITS_COST.value: {
                Ecmsc.LIMITS_COST_MIN.value: cost_min,
                Ecmsc.LIMITS_COST_MAX.value: cost_max,
            }
        }

    def _get_price_precision(self):
        return -decimal.Decimal(f"{self.price_example}").as_tuple().exponent

    def _fix_market_status_precision_with_price(self):
        precision = self._get_price_precision()
        # only patch value when necessary
        if not is_ms_valid(
            self.market_status[Ecmsc.PRECISION.value].get(Ecmsc.PRECISION_AMOUNT.value),
            zero_valid=True
        ):
            self.market_status[Ecmsc.PRECISION.value][Ecmsc.PRECISION_AMOUNT.value] = precision
        if not is_ms_valid(
            self.market_status[Ecmsc.PRECISION.value].get(Ecmsc.PRECISION_PRICE.value),
            zero_valid=True
        ):
            self.market_status[Ecmsc.PRECISION.value][Ecmsc.PRECISION_PRICE.value] = precision

    def _fix_market_status_precision_with_specific(self):
        # binance specific
        pass  # nothing for binance

    def _fix_market_status_limits_with_specific(self):
        market_limit = self.market_status[Ecmsc.LIMITS.value]

        try:
            # binance specific
            if Ecmsic.FILTERS.value in self.market_status_specific:
                filters = self.market_status_specific[Ecmsic.FILTERS.value]

                for filter_dict in filters:
                    if Ecmsic.FILTER_TYPE.value in filter_dict:
                        if filter_dict[Ecmsic.FILTER_TYPE.value] == Ecmsic.PRICE_FILTER.value:
                            if is_ms_valid(float(filter_dict[Ecmsic.MAX_PRICE.value])) and \
                                    not is_ms_valid(
                                        market_limit[Ecmsc.LIMITS_PRICE.value][Ecmsc.LIMITS_PRICE_MAX.value]):
                                market_limit[Ecmsc.LIMITS_PRICE.value][Ecmsc.LIMITS_PRICE_MAX.value] = \
                                    float(filter_dict[Ecmsic.MAX_PRICE.value])

                            if is_ms_valid(float(filter_dict[Ecmsic.MIN_PRICE.value])) and \
                                    not is_ms_valid(
                                        market_limit[Ecmsc.LIMITS_PRICE.value][Ecmsc.LIMITS_PRICE_MIN.value]):
                                market_limit[Ecmsc.LIMITS_PRICE.value][Ecmsc.LIMITS_PRICE_MIN.value] = \
                                    float(filter_dict[Ecmsic.MIN_PRICE.value])
                        elif filter_dict[Ecmsic.FILTER_TYPE.value] == Ecmsic.LOT_SIZE.value:
                            if is_ms_valid(float(filter_dict[Ecmsic.MAX_QTY.value])) and \
                                    not is_ms_valid(
                                        market_limit[Ecmsc.LIMITS_AMOUNT.value][Ecmsc.LIMITS_AMOUNT_MAX.value]):
                                market_limit[Ecmsc.LIMITS_AMOUNT.value][Ecmsc.LIMITS_AMOUNT_MAX.value] = \
                                    float(filter_dict[Ecmsic.MAX_QTY.value])

                            if is_ms_valid(float(filter_dict[Ecmsic.MIN_QTY.value])) and \
                                    not is_ms_valid(
                                        market_limit[Ecmsc.LIMITS_AMOUNT.value][Ecmsc.LIMITS_AMOUNT_MIN.value]):
                                market_limit[Ecmsc.LIMITS_AMOUNT.value][Ecmsc.LIMITS_AMOUNT_MIN.value] = \
                                    float(filter_dict[Ecmsic.MIN_QTY.value])
                calculate_costs(market_limit)
        except Exception:
            pass

File: packages/trading/octobot-trading-ts/src/util/exchangeMarketStatusFixer.ts
import {
  ExchangeConstantsMarketStatusColumns as Ecmsc,
  ExchangeConstantsMarketStatusInfoColumns as Ecmsic,
} from "../enums.js";
import { Logger, getLogger } from "./logger.js";

export type MarketStatus = Record<string, any>;

export function isMsValid(value: unknown, zeroValid = false): boolean {
  let v: number;
  if (typeof value === "string") {
    const parsed = parseFloat(value);
    if (Number.isNaN(parsed)) return false;
    v = parsed;
  } else if (typeof value === "number") {
    v = value;
  } else {
    return false;
  }
  if (Number.isNaN(v)) return false;
  return zeroValid ? v >= 0 : v > 0;
}

export function checkMarketStatusValues(values: Iterable<unknown>, zeroValid = false): boolean {
  for (const v of values) {
    if (!isMsValid(v, zeroValid)) return false;
  }
  return true;
}

export function checkMarketStatusLimits(marketLimit: Record<string, Record<string, unknown>>): boolean {
  for (const key of Object.keys(marketLimit)) {
    if (!checkMarketStatusValues(Object.values(marketLimit[key]))) return false;
  }
  return true;
}

function getMarketsLimit(marketLimit: Record<string, Record<string, any>>) {
  return {
    limitCost: Ecmsc.LIMITS_COST in marketLimit ? marketLimit[Ecmsc.LIMITS_COST] : null,
    limitPrice: Ecmsc.LIMITS_PRICE in marketLimit ? marketLimit[Ecmsc.LIMITS_PRICE] : null,
    limitAmount: Ecmsc.LIMITS_AMOUNT in marketLimit ? marketLimit[Ecmsc.LIMITS_AMOUNT] : null,
  };
}

export function calculateAmounts(marketLimit: Record<string, Record<string, any>>): void {
  const { limitCost, limitPrice, limitAmount } = getMarketsLimit(marketLimit);
  if (!limitAmount || !limitCost || !limitPrice) return;

  if (
    !isMsValid(limitAmount[Ecmsc.LIMITS_AMOUNT_MAX]) &&
    Ecmsc.LIMITS_COST_MAX in limitCost &&
    Ecmsc.LIMITS_PRICE_MAX in limitPrice
  ) {
    if (
      isMsValid(limitCost[Ecmsc.LIMITS_COST_MAX]) &&
      isMsValid(limitPrice[Ecmsc.LIMITS_PRICE_MAX]) &&
      Number(limitPrice[Ecmsc.LIMITS_PRICE_MAX]) > 0
    ) {
      limitAmount[Ecmsc.LIMITS_AMOUNT_MAX] =
        Number(limitCost[Ecmsc.LIMITS_COST_MAX]) / Number(limitPrice[Ecmsc.LIMITS_PRICE_MAX]);
    }
  }

  if (
    !isMsValid(limitAmount[Ecmsc.LIMITS_AMOUNT_MIN]) &&
    Ecmsc.LIMITS_COST_MIN in limitCost &&
    Ecmsc.LIMITS_PRICE_MIN in limitPrice
  ) {
    if (
      isMsValid(limitCost[Ecmsc.LIMITS_COST_MIN]) &&
      isMsValid(limitPrice[Ecmsc.LIMITS_PRICE_MIN]) &&
      Number(limitPrice[Ecmsc.LIMITS_PRICE_MIN]) > 0
    ) {
      limitAmount[Ecmsc.LIMITS_AMOUNT_MIN] =
        Number(limitCost[Ecmsc.LIMITS_COST_MIN]) / Number(limitPrice[Ecmsc.LIMITS_PRICE_MIN]);
    }
  }
}

export function calculateCosts(marketLimit: Record<string, Record<string, any>>): void {
  const { limitCost, limitPrice, limitAmount } = getMarketsLimit(marketLimit);
  if (!limitAmount || !limitCost || !limitPrice) return;

  if (
    !isMsValid(limitCost[Ecmsc.LIMITS_COST_MAX]) &&
    Ecmsc.LIMITS_AMOUNT_MAX in limitAmount &&
    Ecmsc.LIMITS_PRICE_MAX in limitPrice
  ) {
    if (isMsValid(limitAmount[Ecmsc.LIMITS_AMOUNT_MAX]) && isMsValid(limitPrice[Ecmsc.LIMITS_PRICE_MAX])) {
      limitCost[Ecmsc.LIMITS_COST_MAX] =
        Number(limitAmount[Ecmsc.LIMITS_AMOUNT_MAX]) * Number(limitPrice[Ecmsc.LIMITS_PRICE_MAX]);
    }
  }

  if (
    !isMsValid(limitCost[Ecmsc.LIMITS_COST_MIN]) &&
    Ecmsc.LIMITS_AMOUNT_MIN in limitAmount &&
    Ecmsc.LIMITS_PRICE_MIN in limitPrice
  ) {
    if (isMsValid(limitAmount[Ecmsc.LIMITS_AMOUNT_MIN]) && isMsValid(limitPrice[Ecmsc.LIMITS_PRICE_MIN])) {
      limitCost[Ecmsc.LIMITS_COST_MIN] =
        Number(limitAmount[Ecmsc.LIMITS_AMOUNT_MIN]) * Number(limitPrice[Ecmsc.LIMITS_PRICE_MIN]);
    }
  }
}

export function updatePrices(marketLimit: Record<string, Record<string, any>>): void {
  const { limitPrice } = getMarketsLimit(marketLimit);
  if (!limitPrice) return;
  if (!isMsValid(limitPrice[Ecmsc.LIMITS_PRICE_MAX])) limitPrice[Ecmsc.LIMITS_PRICE_MAX] = null;
  if (!isMsValid(limitPrice[Ecmsc.LIMITS_PRICE_MIN])) limitPrice[Ecmsc.LIMITS_PRICE_MIN] = null;
}

export function fixMarketStatusLimitsFromCurrentData(marketLimit: Record<string, Record<string, any>>): void {
  if (!checkMarketStatusValues(Object.values(marketLimit[Ecmsc.LIMITS_COST] || {}))) {
    calculateCosts(marketLimit);
  }
  if (!checkMarketStatusValues(Object.values(marketLimit[Ecmsc.LIMITS_AMOUNT] || {}))) {
    calculateAmounts(marketLimit);
  }
  if (!checkMarketStatusValues(Object.values(marketLimit[Ecmsc.LIMITS_PRICE] || {}))) {
    updatePrices(marketLimit);
  }
  if (!isMsValid(marketLimit[Ecmsc.LIMITS_COST]?.[Ecmsc.LIMITS_COST_MIN])) {
    if (!marketLimit[Ecmsc.LIMITS_COST]) marketLimit[Ecmsc.LIMITS_COST] = {};
    marketLimit[Ecmsc.LIMITS_COST][Ecmsc.LIMITS_COST_MIN] = 0;
  }
}

export class ExchangeMarketStatusFixer {
  static readonly LIMIT_PRICE_MULTIPLIER = 1000;
  static readonly LIMIT_COST_MULTIPLIER = 1;

  static readonly LIMIT_AMOUNT_MAX_SUP_ATTENUATION = 8;
  static readonly LIMIT_AMOUNT_MAX_MINUS_3_ATTENUATION = 8;
  static readonly LIMIT_AMOUNT_MIN_ATTENUATION = 3;
  static readonly LIMIT_AMOUNT_MIN_SUP_ATTENUATION = 1;

  protected logger: Logger;
  marketStatus: MarketStatus;
  priceExample: number | null;
  marketStatusSpecific: Record<string, any> | null;

  constructor(marketStatus: MarketStatus, priceExample: number | null = null) {
    this.logger = getLogger(this.constructor.name);
    this.marketStatus = marketStatus;
    this.priceExample = priceExample;
    this.marketStatusSpecific = (Ecmsc.INFO in this.marketStatus)
      ? (this.marketStatus[Ecmsc.INFO] as Record<string, any>)
      : null;

    this.fixTyping();
    this.fixMarketStatusPrecision();
    this.fixMarketStatusLimits();
  }

  protected convertValuesToFloat(
    element: Record<string, any>,
    parentKeys: string[],
    keyWhitelist: string[],
  ): void {
    let cur: Record<string, any> | undefined = element;
    for (const k of parentKeys) {
      if (!cur || !(k in cur)) return;
      cur = cur[k] as Record<string, any>;
    }
    if (!cur) return;
    for (const [k, v] of Object.entries(cur)) {
      if (keyWhitelist.includes(k) && typeof v === "string") {
        const parsed = parseFloat(v);
        if (!Number.isNaN(parsed)) cur[k] = parsed;
        else this.logger.debug(`Impossible to convert ${v} to float in ${k} market status`);
      }
    }
  }

  protected fixTyping(): void {
    this.convertValuesToFloat(
      this.marketStatus,
      [Ecmsc.PRECISION],
      [Ecmsc.PRECISION_AMOUNT, Ecmsc.PRECISION_PRICE],
    );
    this.convertValuesToFloat(
      this.marketStatus,
      [Ecmsc.LIMITS, Ecmsc.LIMITS_COST],
      [Ecmsc.LIMITS_COST_MAX, Ecmsc.LIMITS_COST_MIN],
    );
    this.convertValuesToFloat(
      this.marketStatus,
      [Ecmsc.LIMITS, Ecmsc.LIMITS_AMOUNT],
      [Ecmsc.LIMITS_AMOUNT_MAX, Ecmsc.LIMITS_AMOUNT_MIN],
    );
    this.convertValuesToFloat(
      this.marketStatus,
      [Ecmsc.LIMITS, Ecmsc.LIMITS_PRICE],
      [Ecmsc.LIMITS_PRICE_MAX, Ecmsc.LIMITS_PRICE_MIN],
    );
  }

  protected fixMarketStatusPrecision(): void {
    if (!(Ecmsc.PRECISION in this.marketStatus)) {
      this.marketStatus[Ecmsc.PRECISION] = {
        [Ecmsc.PRECISION_AMOUNT]: null,
        [Ecmsc.PRECISION_PRICE]: null,
      };
    }
    const marketPrecision = this.marketStatus[Ecmsc.PRECISION];
    if (!checkMarketStatusValues(
      [marketPrecision[Ecmsc.PRECISION_AMOUNT], marketPrecision[Ecmsc.PRECISION_PRICE]],
      true,
    )) {
      if (this.priceExample !== null) {
        this.fixMarketStatusPrecisionWithPrice();
      } else if (this.marketStatusSpecific) {
        this.fixMarketStatusPrecisionWithSpecific();
      }
    }
  }

  protected fixMarketStatusLimits(): void {
    if (!(Ecmsc.LIMITS in this.marketStatus)) this.marketStatus[Ecmsc.LIMITS] = {};
    const marketLimit = this.marketStatus[Ecmsc.LIMITS];
    if (!(Ecmsc.LIMITS_COST in marketLimit)) {
      marketLimit[Ecmsc.LIMITS_COST] = { [Ecmsc.LIMITS_COST_MAX]: null, [Ecmsc.LIMITS_COST_MIN]: null };
    }
    if (!(Ecmsc.LIMITS_AMOUNT in marketLimit)) {
      marketLimit[Ecmsc.LIMITS_AMOUNT] = { [Ecmsc.LIMITS_AMOUNT_MAX]: null, [Ecmsc.LIMITS_AMOUNT_MIN]: null };
    }
    if (!(Ecmsc.LIMITS_PRICE in marketLimit)) {
      marketLimit[Ecmsc.LIMITS_PRICE] = { [Ecmsc.LIMITS_PRICE_MAX]: null, [Ecmsc.LIMITS_PRICE_MIN]: null };
    }
    if (!checkMarketStatusLimits(marketLimit)) {
      fixMarketStatusLimitsFromCurrentData(marketLimit);
      if (this.marketStatusSpecific && !checkMarketStatusLimits(marketLimit)) {
        this.fixMarketStatusLimitsWithSpecific();
      }
      if (this.priceExample !== null && !checkMarketStatusLimits(marketLimit)) {
        this.fixMarketStatusLimitsWithPrice();
      }
    }
  }

  protected calculateAmount(): { amountMin: number; amountMax: number } {
    const px = this.priceExample as number;
    const amountLogPrice = Math.log10(px);
    if (amountLogPrice >= 0) {
      return {
        amountMin: 10 ** (ExchangeMarketStatusFixer.LIMIT_AMOUNT_MIN_SUP_ATTENUATION - amountLogPrice),
        amountMax: 10 ** (ExchangeMarketStatusFixer.LIMIT_AMOUNT_MAX_SUP_ATTENUATION - amountLogPrice),
      };
    }
    return {
      amountMin: 10 ** -(amountLogPrice + ExchangeMarketStatusFixer.LIMIT_AMOUNT_MIN_ATTENUATION),
      amountMax: 10 ** (-amountLogPrice + ExchangeMarketStatusFixer.LIMIT_AMOUNT_MAX_MINUS_3_ATTENUATION),
    };
  }

  protected fixMarketStatusLimitsWithPrice(): void {
    const px = this.priceExample as number;
    const { amountMin: candidateAmountMin, amountMax: candidateAmountMax } = this.calculateAmount();
    const limitsAmount = this.marketStatus[Ecmsc.LIMITS][Ecmsc.LIMITS_AMOUNT];
    let amountMin = limitsAmount[Ecmsc.LIMITS_AMOUNT_MIN];
    if (!isMsValid(amountMin, true)) amountMin = candidateAmountMin;
    let amountMax = limitsAmount[Ecmsc.LIMITS_AMOUNT_MAX];
    if (!isMsValid(amountMax, false)) amountMax = candidateAmountMax;

    const limitsPrice = this.marketStatus[Ecmsc.LIMITS][Ecmsc.LIMITS_PRICE];
    let priceMin = limitsPrice[Ecmsc.LIMITS_PRICE_MIN];
    if (!isMsValid(priceMin, true)) priceMin = px / ExchangeMarketStatusFixer.LIMIT_PRICE_MULTIPLIER;
    let priceMax = limitsPrice[Ecmsc.LIMITS_PRICE_MAX];
    if (!isMsValid(priceMax, false)) priceMax = px * ExchangeMarketStatusFixer.LIMIT_PRICE_MULTIPLIER;

    const limitCost = this.marketStatus[Ecmsc.LIMITS][Ecmsc.LIMITS_COST] || {};
    let costMin = limitCost[Ecmsc.LIMITS_COST_MIN];
    if (!isMsValid(costMin, false)) {
      let updatedCost = false;
      if (
        isMsValid(limitsPrice[Ecmsc.LIMITS_PRICE_MIN]) &&
        isMsValid(limitsAmount[Ecmsc.LIMITS_AMOUNT_MIN])
      ) {
        const candidateCostMin = Number(limitsPrice[Ecmsc.LIMITS_PRICE_MIN]) * Number(limitsAmount[Ecmsc.LIMITS_AMOUNT_MIN]);
        if (isMsValid(candidateCostMin, false)) {
          costMin = candidateCostMin;
          updatedCost = true;
        }
      }
      if (!updatedCost) costMin = 0;
    }
    let costMax = limitCost[Ecmsc.LIMITS_COST_MAX];
    if (!isMsValid(costMax, false)) costMax = priceMax * amountMax;

    this.marketStatus[Ecmsc.LIMITS] = {
      [Ecmsc.LIMITS_AMOUNT]: { [Ecmsc.LIMITS_AMOUNT_MIN]: amountMin, [Ecmsc.LIMITS_AMOUNT_MAX]: amountMax },
      [Ecmsc.LIMITS_PRICE]: { [Ecmsc.LIMITS_PRICE_MIN]: priceMin, [Ecmsc.LIMITS_PRICE_MAX]: priceMax },
      [Ecmsc.LIMITS_COST]: { [Ecmsc.LIMITS_COST_MIN]: costMin, [Ecmsc.LIMITS_COST_MAX]: costMax },
    };
  }

  protected getPricePrecision(): number {
    const s = String(this.priceExample);
    if (!s.includes(".")) return 0;
    return s.split(".")[1].length;
  }

  protected fixMarketStatusPrecisionWithPrice(): void {
    const precision = this.getPricePrecision();
    const p = this.marketStatus[Ecmsc.PRECISION];
    if (!isMsValid(p[Ecmsc.PRECISION_AMOUNT], true)) p[Ecmsc.PRECISION_AMOUNT] = precision;
    if (!isMsValid(p[Ecmsc.PRECISION_PRICE], true)) p[Ecmsc.PRECISION_PRICE] = precision;
  }

  protected fixMarketStatusPrecisionWithSpecific(): void {
    // binance specific — no-op like Python.
  }

  protected fixMarketStatusLimitsWithSpecific(): void {
    const marketLimit = this.marketStatus[Ecmsc.LIMITS];
    try {
      if (this.marketStatusSpecific && Ecmsic.FILTERS in this.marketStatusSpecific) {
        const filters = this.marketStatusSpecific[Ecmsic.FILTERS] as Array<Record<string, any>>;
        for (const f of filters) {
          if (!(Ecmsic.FILTER_TYPE in f)) continue;
          if (f[Ecmsic.FILTER_TYPE] === Ecmsic.PRICE_FILTER) {
            const maxPrice = parseFloat(f[Ecmsic.MAX_PRICE]);
            if (
              isMsValid(maxPrice) &&
              !isMsValid(marketLimit[Ecmsc.LIMITS_PRICE][Ecmsc.LIMITS_PRICE_MAX])
            ) {
              marketLimit[Ecmsc.LIMITS_PRICE][Ecmsc.LIMITS_PRICE_MAX] = maxPrice;
            }
            const minPrice = parseFloat(f[Ecmsic.MIN_PRICE]);
            if (
              isMsValid(minPrice) &&
              !isMsValid(marketLimit[Ecmsc.LIMITS_PRICE][Ecmsc.LIMITS_PRICE_MIN])
            ) {
              marketLimit[Ecmsc.LIMITS_PRICE][Ecmsc.LIMITS_PRICE_MIN] = minPrice;
            }
          } else if (f[Ecmsic.FILTER_TYPE] === Ecmsic.LOT_SIZE) {
            const maxQty = parseFloat(f[Ecmsic.MAX_QTY]);
            if (
              isMsValid(maxQty) &&
              !isMsValid(marketLimit[Ecmsc.LIMITS_AMOUNT][Ecmsc.LIMITS_AMOUNT_MAX])
            ) {
              marketLimit[Ecmsc.LIMITS_AMOUNT][Ecmsc.LIMITS_AMOUNT_MAX] = maxQty;
            }
            const minQty = parseFloat(f[Ecmsic.MIN_QTY]);
            if (
              isMsValid(minQty) &&
              !isMsValid(marketLimit[Ecmsc.LIMITS_AMOUNT][Ecmsc.LIMITS_AMOUNT_MIN])
            ) {
              marketLimit[Ecmsc.LIMITS_AMOUNT][Ecmsc.LIMITS_AMOUNT_MIN] = minQty;
            }
          }
        }
        calculateCosts(marketLimit);
      }
    } catch {
      // mirror Python's broad-catch silence
    }
  }
}
